contracts/utils/cryptography/MultiSignerERC7913.sol

```diff
@@ -0,0 +1,246 @@
+
+    /// @dev The `threshold` is unreachable given the number of `signers`.
+    error MultiSignerERC7913UnreachableThreshold(uint256 signers, uint256 threshold);
+
+    /**
+     * @dev Returns the set of authorized signers. Prefer {_signers} for internal use.
+     *
+     * WARNING: This operation copies the entire signers set to memory, which can be expensive. This is designed
+     * for view accessors queried without gas fees. Using it in state-changing functions may become uncallable
+     * if the signers set grows too large.
+     */
+    function signers() public view virtual returns (bytes[] memory) {
+        return _signers().values();
+    }
+
+    /// @dev Returns whether the `signer` is an authorized signer.
+    function isSigner(bytes memory signer) public view virtual returns (bool) {
+        return _signers().contains(signer);
+    }
+
+    /// @dev Returns the minimum number of signers required to approve a multisignature operation.
+    function threshold() public view virtual returns (uint256) {
+        return _threshold;
+    }
+
+    /// @dev Returns the set of authorized signers.
+    function _signers() internal view virtual returns (EnumerableSetExtended.BytesSet storage) {
+        return _signersSet;
+    }
+
+    /**
+     * @dev Adds the `newSigners` to those allowed to sign on behalf of this contract.
+     * Internal version without access control.
+     *
+     * Requirements:
+     *
+     * * Each of `newSigners` must be at least 20 bytes long. Reverts with {MultiSignerERC7913InvalidSigner} if not.
+     * * Each of `newSigners` must not be authorized. See {isSigner}. Reverts with {MultiSignerERC7913AlreadyExists} if so.
+     */
+    function _addSigners(bytes[] memory newSigners) internal virtual {
+        uint256 newSignersLength = newSigners.length;
+        for (uint256 i = 0; i < newSignersLength; i++) {
+            bytes memory signer = newSigners[i];
+            require(signer.length >= 20, MultiSignerERC7913InvalidSigner(signer));
+            require(_signers().add(signer), MultiSignerERC7913AlreadyExists(signer));
+        }
+        emit ERC7913SignersAdded(newSigners);
+    }
+
+    /**
+     * @dev Removes the `oldSigners` from the authorized signers. Internal version without access control.
+     *
+     * Requirements:
+     *
+     * * Each of `oldSigners` must be authorized. See {isSigner}. Otherwise {MultiSignerERC7913NonexistentSigner} is thrown.
+     * * See {_validateReachableThreshold} for the threshold validation.
+     */
+    function _removeSigners(bytes[] memory oldSigners) internal virtual {
+        uint256 oldSignersLength = oldSigners.length;
+        for (uint256 i = 0; i < oldSignersLength; i++) {
+            bytes memory signer = oldSigners[i];
+            require(_signers().remove(signer), MultiSignerERC7913NonexistentSigner(signer));
+        }
+        _validateReachableThreshold();
+        emit ERC7913SignersRemoved(oldSigners);
+    }
```

Comment:
what is the rational for having internal functions that add in batch, vs function that add one and can be called multiple time (like we have in AccessControl).

For the `_removeSigners` I see it avoids computing `_validateReachableThreshold` multiple times, but I'd still go for the simple version that doesn't deal with arrays.